File: api/src/test/java/com/api/app/service/basket/BasketServiceImplTest.java
package com.api.app.service.basket;

import com.api.app.common.exception.ApiError;
import com.api.app.common.exception.ApiException;
import com.api.app.dto.request.basket.BasketAddRequest;
import com.api.app.dto.request.basket.BasketModifyRequest;
import com.api.app.dto.response.basket.BasketResponse;
import com.api.app.entity.BasketBase;
import com.api.app.entity.MemberBase;
import com.api.app.repository.basket.BasketBaseMapper;
import com.api.app.repository.basket.BasketBaseTrxMapper;
import com.api.app.repository.member.MemberBaseMapper;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.*;

/**
 * @author system
 * @version 1.0
 * @since 2025-10-29
 */
@ExtendWith(MockitoExtension.class)
class BasketServiceImplTest {

    @InjectMocks
    private BasketServiceImpl basketService;

    @Mock
    private BasketBaseMapper basketBaseMapper;

    @Mock
    private BasketBaseTrxMapper basketBaseTrxMapper;

    @Mock
    private MemberBaseMapper memberBaseMapper;

    @Test
    @DisplayName("장바구니 목록 조회 성공")
    void getBasketList_Success() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);
        member.setEmail(email);

        List<BasketResponse> basketList = new ArrayList<>();
        BasketResponse basket1 = new BasketResponse();
        basket1.setBasketNo("000000000000001");
        basket1.setGoodsNo("G00000000000001");
        basket1.setItemNo("001");
        basket1.setQuantity(2L);
        basketList.add(basket1);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketListByMemberNo(memberNo)).willReturn(basketList);

        // when
        List<BasketResponse> result = basketService.getBasketList(email);

        // then
        assertThat(result).hasSize(1);
        assertThat(result.get(0).getBasketNo()).isEqualTo("000000000000001");
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(1)).selectBasketListByMemberNo(memberNo);
    }

    @Test
    @DisplayName("장바구니 추가 성공 - 새로운 상품")
    void addBasket_Success_NewItem() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        String basketNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketAddRequest request = new BasketAddRequest();
        request.setGoodsNo("G00000000000001");
        request.setItemNo("001");
        request.setQuantity(1L);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByMemberAndItem(memberNo, request.getGoodsNo(), request.getItemNo()))
                .willReturn(null);
        given(basketBaseTrxMapper.generateBasketNo()).willReturn(basketNo);
        given(basketBaseTrxMapper.insertBasketBase(any(BasketBase.class))).willReturn(1);

        // when
        String result = basketService.addBasket(email, request);

        // then
        assertThat(result).isEqualTo(basketNo);
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(1)).selectBasketBaseByMemberAndItem(memberNo, request.getGoodsNo(), request.getItemNo());
        verify(basketBaseTrxMapper, times(1)).generateBasketNo();
        verify(basketBaseTrxMapper, times(1)).insertBasketBase(any(BasketBase.class));
    }

    @Test
    @DisplayName("장바구니 추가 성공 - 기존 상품 수량 증가")
    void addBasket_Success_UpdateQuantity() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        String basketNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketAddRequest request = new BasketAddRequest();
        request.setGoodsNo("G00000000000001");
        request.setItemNo("001");
        request.setQuantity(2L);

        BasketBase existingBasket = new BasketBase();
        existingBasket.setBasketNo(basketNo);
        existingBasket.setMemberNo(memberNo);
        existingBasket.setGoodsNo(request.getGoodsNo());
        existingBasket.setItemNo(request.getItemNo());
        existingBasket.setQuantity(3L);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByMemberAndItem(memberNo, request.getGoodsNo(), request.getItemNo()))
                .willReturn(existingBasket);
        given(basketBaseTrxMapper.updateBasketBase(any(BasketBase.class))).willReturn(1);

        // when
        String result = basketService.addBasket(email, request);

        // then
        assertThat(result).isEqualTo(basketNo);
        assertThat(existingBasket.getQuantity()).isEqualTo(5L); // 3 + 2
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseTrxMapper, times(1)).updateBasketBase(any(BasketBase.class));
        verify(basketBaseTrxMapper, never()).insertBasketBase(any(BasketBase.class));
    }

    @Test
    @DisplayName("장바구니 수정 성공")
    void modifyBasket_Success() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        String basketNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketBase basketBase = new BasketBase();
        basketBase.setBasketNo(basketNo);
        basketBase.setMemberNo(memberNo);
        basketBase.setQuantity(1L);

        BasketModifyRequest request = new BasketModifyRequest();
        request.setQuantity(5L);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByBasketNo(basketNo)).willReturn(basketBase);
        given(basketBaseTrxMapper.updateBasketBase(any(BasketBase.class))).willReturn(1);

        // when
        basketService.modifyBasket(email, basketNo, request);

        // then
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(1)).selectBasketBaseByBasketNo(basketNo);
        verify(basketBaseTrxMapper, times(1)).updateBasketBase(any(BasketBase.class));
    }

    @Test
    @DisplayName("장바구니 수정 실패 - 장바구니 없음")
    void modifyBasket_Fail_BasketNotFound() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        String basketNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketModifyRequest request = new BasketModifyRequest();
        request.setQuantity(5L);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByBasketNo(basketNo)).willReturn(null);

        // when & then
        assertThatThrownBy(() -> basketService.modifyBasket(email, basketNo, request))
                .isInstanceOf(ApiException.class)
                .hasFieldOrPropertyWithValue("apiError", ApiError.DATA_NOT_FOUND);

        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(1)).selectBasketBaseByBasketNo(basketNo);
        verify(basketBaseTrxMapper, never()).updateBasketBase(any(BasketBase.class));
    }

    @Test
    @DisplayName("장바구니 수정 실패 - 권한 없음")
    void modifyBasket_Fail_Forbidden() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        String basketNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketBase basketBase = new BasketBase();
        basketBase.setBasketNo(basketNo);
        basketBase.setMemberNo("000000000000002"); // 다른 회원의 장바구니

        BasketModifyRequest request = new BasketModifyRequest();
        request.setQuantity(5L);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByBasketNo(basketNo)).willReturn(basketBase);

        // when & then
        assertThatThrownBy(() -> basketService.modifyBasket(email, basketNo, request))
                .isInstanceOf(ApiException.class)
                .hasFieldOrPropertyWithValue("apiError", ApiError.FORBIDDEN);

        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(1)).selectBasketBaseByBasketNo(basketNo);
        verify(basketBaseTrxMapper, never()).updateBasketBase(any(BasketBase.class));
    }

    @Test
    @DisplayName("장바구니 삭제 성공")
    void deleteBasket_Success() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        String basketNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketBase basketBase = new BasketBase();
        basketBase.setBasketNo(basketNo);
        basketBase.setMemberNo(memberNo);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByBasketNo(basketNo)).willReturn(basketBase);
        given(basketBaseTrxMapper.deleteBasketBase(basketNo)).willReturn(1);

        // when
        basketService.deleteBasket(email, basketNo);

        // then
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(1)).selectBasketBaseByBasketNo(basketNo);
        verify(basketBaseTrxMapper, times(1)).deleteBasketBase(basketNo);
    }

    @Test
    @DisplayName("장바구니 여러 개 삭제 성공")
    void deleteBaskets_Success() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";
        List<String> basketNos = Arrays.asList("000000000000001", "000000000000002");

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        BasketBase basket1 = new BasketBase();
        basket1.setBasketNo("000000000000001");
        basket1.setMemberNo(memberNo);

        BasketBase basket2 = new BasketBase();
        basket2.setBasketNo("000000000000002");
        basket2.setMemberNo(memberNo);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseMapper.selectBasketBaseByBasketNo("000000000000001")).willReturn(basket1);
        given(basketBaseMapper.selectBasketBaseByBasketNo("000000000000002")).willReturn(basket2);
        given(basketBaseTrxMapper.deleteBasketBaseByNos(basketNos)).willReturn(2);

        // when
        basketService.deleteBaskets(email, basketNos);

        // then
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, times(2)).selectBasketBaseByBasketNo(anyString());
        verify(basketBaseTrxMapper, times(1)).deleteBasketBaseByNos(basketNos);
    }

    @Test
    @DisplayName("장바구니 전체 삭제 성공")
    void deleteAllBaskets_Success() {
        // given
        String email = "test@example.com";
        String memberNo = "000000000000001";

        MemberBase member = new MemberBase();
        member.setMemberNo(memberNo);

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(member);
        given(basketBaseTrxMapper.deleteBasketBaseByMemberNo(memberNo)).willReturn(5);

        // when
        basketService.deleteAllBaskets(email);

        // then
        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseTrxMapper, times(1)).deleteBasketBaseByMemberNo(memberNo);
    }

    @Test
    @DisplayName("장바구니 조회 실패 - 회원 정보 없음")
    void getBasketList_Fail_MemberNotFound() {
        // given
        String email = "test@example.com";

        given(memberBaseMapper.selectMemberBaseByEmail(email)).willReturn(null);

        // when & then
        assertThatThrownBy(() -> basketService.getBasketList(email))
                .isInstanceOf(ApiException.class)
                .hasFieldOrPropertyWithValue("apiError", ApiError.DATA_NOT_FOUND);

        verify(memberBaseMapper, times(1)).selectMemberBaseByEmail(email);
        verify(basketBaseMapper, never()).selectBasketListByMemberNo(anyString());
    }
}

File: api/src/test/kotlin/com/api/app/service/basket/BasketServiceImplTest.kt
package com.api.app.service.basket

import com.api.app.common.exception.ApiError
import com.api.app.common.exception.ApiException
import com.api.app.dto.request.basket.BasketAddRequest
import com.api.app.dto.request.basket.BasketModifyRequest
import com.api.app.emum.PRD001
import com.api.app.entity.BasketBase
import com.api.app.entity.MemberBase
import com.api.app.repository.rodb.basket.BasketBaseRepository
import com.api.app.repository.rodb.basket.BasketProjection
import com.api.app.repository.rodb.member.MemberBaseRepository
import com.api.app.repository.rwdb.basket.BasketBaseTrxRepository
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import java.time.LocalDateTime
import java.util.Optional

@ExtendWith(MockitoExtension::class)
class BasketServiceImplTest {

    @InjectMocks
    private lateinit var basketService: BasketServiceImpl

    @Mock
    private lateinit var basketBaseRepository: BasketBaseRepository

    @Mock
    private lateinit var basketBaseTrxRepository: BasketBaseTrxRepository

    @Mock
    private lateinit var memberBaseRepository: MemberBaseRepository

    @Test
    @DisplayName("장바구니 목록 조회 성공")
    fun getBasketListSuccess() {
        val email = "test@example.com"
        val member = createMember(email = email)
        val projection = createProjection(basketNo = "B001", memberNo = member.memberNo)
        given(memberBaseRepository.findByEmail(email)).willReturn(member)
        given(basketBaseRepository.selectBasketListByMemberNo(member.memberNo)).willReturn(listOf(projection))

        val result = basketService.getBasketList(email)

        assertThat(result).hasSize(1)
        assertThat(result.single().basketNo).isEqualTo("B001")
        assertThat(result.single().goodsStatusCode).isEqualTo(PRD001.ON_SALE.code)
        verify(basketBaseRepository, times(1)).selectBasketListByMemberNo(member.memberNo)
    }

    @Test
    @DisplayName("장바구니 목록 조회 실패 - 회원 없음")
    fun getBasketListFailWhenMemberMissing() {
        given(memberBaseRepository.findByEmail("missing@example.com")).willReturn(null)

        assertThatThrownBy { basketService.getBasketList("missing@example.com") }
            .isInstanceOf(ApiException::class.java)
            .hasFieldOrPropertyWithValue("apiError", ApiError.DATA_NOT_FOUND)
    }

    @Test
    @DisplayName("장바구니 추가 성공 - 신규 상품")
    fun addBasketSuccessForNewBasket() {
        val email = "test@example.com"
        val member = createMember(email = email)
        val request = BasketAddRequest(goodsNo = "G001", itemNo = "001", quantity = 2L)
        given(memberBaseRepository.findByEmail(email)).willReturn(member)
        given(
            basketBaseRepository.findByMemberNoAndGoodsNoAndItemNoAndIsOrderFalse(
                member.memberNo,
                request.goodsNo,
                request.itemNo
            )
        ).willReturn(null)
        given(basketBaseTrxRepository.save(any(BasketBase::class.java))).willAnswer { invocation ->
            (invocation.arguments[0] as BasketBase).apply { basketNo = "B001" }
        }

        val basketNo = basketService.addBasket(email, request)

        val captor = ArgumentCaptor.forClass(BasketBase::class.java)
        verify(basketBaseTrxRepository).save(captor.capture())
        assertThat(basketNo).isEqualTo("B001")
        assertThat(captor.value.memberNo).isEqualTo(member.memberNo)
        assertThat(captor.value.isOrder).isFalse()
        assertThat(captor.value.quantity).isEqualTo(2L)
    }

    @Test
    @DisplayName("장바구니 추가 성공 - 기존 수량 증가")
    fun addBasketSuccessForExistingBasket() {
        val email = "test@example.com"
        val member = createMember(email = email)
        val request = BasketAddRequest(goodsNo = "G001", itemNo = "001", quantity = 2L)
        val existing = BasketBase().apply {
            basketNo = "B001"
            memberNo = member.memberNo
            goodsNo = request.goodsNo
            itemNo = request.itemNo
            quantity = 3L
        }

        given(memberBaseRepository.findByEmail(email)).willReturn(member)
        given(
            basketBaseRepository.findByMemberNoAndGoodsNoAndItemNoAndIsOrderFalse(
                member.memberNo,
                request.goodsNo,
                request.itemNo
            )
        ).willReturn(existing)

        val basketNo = basketService.addBasket(email, request)

        assertThat(basketNo).isEqualTo("B001")
        assertThat(existing.quantity).isEqualTo(5L)
        verify(basketBaseTrxRepository, times(1)).save(existing)
    }

    @Test
    @DisplayName("장바구니 수정 성공")
    fun modifyBasketSuccess() {
        val email = "test@example.com"
        val member = createMember(email = email)
        val basket = BasketBase().apply {
            basketNo = "B001"
            memberNo = member.memberNo
            goodsNo = "G001"
            itemNo = "001"
            quantity = 1L
        }
        val request = BasketModifyRequest(goodsNo = "G002", itemNo = "002", quantity = 5L)
        given(memberBaseRepository.findByEmail(email)).willReturn(member)
        given(basketBaseRepository.findById("B001")).willReturn(Optional.of(basket))

        basketService.modifyBasket(email, "B001", request)

        assertThat(basket.goodsNo).isEqualTo("G002")
        assertThat(basket.itemNo).isEqualTo("002")
        assertThat(basket.quantity).isEqualTo(5L)
        verify(basketBaseTrxRepository).save(basket)
    }

    @Test
    @DisplayName("장바구니 수정 실패 - 장바구니 없음")
    fun modifyBasketFailWhenBasketMissing() {
        val member = createMember(email = "test@example.com")
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)
        given(basketBaseRepository.findById("B001")).willReturn(Optional.empty())

        assertThatThrownBy {
            basketService.modifyBasket(member.email, "B001", BasketModifyRequest(quantity = 1L))
        }
            .isInstanceOf(ApiException::class.java)
            .hasFieldOrPropertyWithValue("apiError", ApiError.DATA_NOT_FOUND)
    }

    @Test
    @DisplayName("장바구니 수정 실패 - 권한 없음")
    fun modifyBasketFailWhenForbidden() {
        val member = createMember(email = "test@example.com")
        val basket = BasketBase().apply {
            basketNo = "B001"
            memberNo = "OTHER"
        }
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)
        given(basketBaseRepository.findById("B001")).willReturn(Optional.of(basket))

        assertThatThrownBy {
            basketService.modifyBasket(member.email, "B001", BasketModifyRequest(quantity = 1L))
        }
            .isInstanceOf(ApiException::class.java)
            .hasFieldOrPropertyWithValue("apiError", ApiError.FORBIDDEN)

        verify(basketBaseTrxRepository, never()).save(any(BasketBase::class.java))
    }

    @Test
    @DisplayName("장바구니 삭제 성공")
    fun deleteBasketSuccess() {
        val member = createMember(email = "test@example.com")
        val basket = BasketBase().apply {
            basketNo = "B001"
            memberNo = member.memberNo
        }
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)
        given(basketBaseRepository.findById("B001")).willReturn(Optional.of(basket))

        basketService.deleteBasket(member.email, "B001")

        verify(basketBaseTrxRepository).deleteById("B001")
    }

    @Test
    @DisplayName("장바구니 삭제 실패 - 권한 없음")
    fun deleteBasketFailWhenForbidden() {
        val member = createMember(email = "test@example.com")
        val basket = BasketBase().apply {
            basketNo = "B001"
            memberNo = "OTHER"
        }
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)
        given(basketBaseRepository.findById("B001")).willReturn(Optional.of(basket))

        assertThatThrownBy { basketService.deleteBasket(member.email, "B001") }
            .isInstanceOf(ApiException::class.java)
            .hasFieldOrPropertyWithValue("apiError", ApiError.FORBIDDEN)

        verify(basketBaseTrxRepository, never()).deleteById("B001")
    }

    @Test
    @DisplayName("장바구니 여러 개 삭제 성공")
    fun deleteBasketsSuccess() {
        val member = createMember(email = "test@example.com")
        val basketNos = listOf("B001", "B002")
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)
        basketNos.forEach { basketNo ->
            given(basketBaseRepository.findById(basketNo)).willReturn(
                Optional.of(BasketBase().apply {
                    this.basketNo = basketNo
                    memberNo = member.memberNo
                })
            )
        }

        basketService.deleteBaskets(member.email, basketNos)

        verify(basketBaseTrxRepository).deleteByBasketNoIn(basketNos)
    }

    @Test
    @DisplayName("장바구니 여러 개 삭제 실패 - 비어 있는 요청")
    fun deleteBasketsFailWhenEmpty() {
        assertThatThrownBy { basketService.deleteBaskets("test@example.com", emptyList()) }
            .isInstanceOf(ApiException::class.java)
            .hasFieldOrPropertyWithValue("apiError", ApiError.INVALID_PARAMETER)
    }

    @Test
    @DisplayName("장바구니 여러 개 삭제 실패 - 일부 장바구니 없음")
    fun deleteBasketsFailWhenBasketMissing() {
        val member = createMember(email = "test@example.com")
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)
        given(basketBaseRepository.findById("B001")).willReturn(Optional.empty())

        assertThatThrownBy { basketService.deleteBaskets(member.email, listOf("B001")) }
            .isInstanceOf(ApiException::class.java)
            .hasFieldOrPropertyWithValue("apiError", ApiError.DATA_NOT_FOUND)
    }

    @Test
    @DisplayName("장바구니 전체 삭제 성공")
    fun deleteAllBasketsSuccess() {
        val member = createMember(email = "test@example.com")
        given(memberBaseRepository.findByEmail(member.email)).willReturn(member)

        basketService.deleteAllBaskets(member.email)

        verify(basketBaseTrxRepository).deleteByMemberNo(member.memberNo)
    }

    private fun createMember(
        memberNo: String = "M001",
        email: String = "test@example.com"
    ) = MemberBase().apply {
        this.memberNo = memberNo
        this.email = email
    }

    private fun createProjection(
        basketNo: String,
        memberNo: String = "M001"
    ): BasketProjection {
        val projection = mock(BasketProjection::class.java)
        doReturn(basketNo).`when`(projection).getBasketNo()
        doReturn(memberNo).`when`(projection).getMemberNo()
        doReturn("G001").`when`(projection).getGoodsNo()
        doReturn("테스트상품").`when`(projection).getGoodsName()
        doReturn(PRD001.ON_SALE.code).`when`(projection).getGoodsStatusCode()
        doReturn("https://cdn.example.com/goods.jpg").`when`(projection).getGoodsMainImageUrl()
        doReturn(10000L).`when`(projection).getSalePrice()
        doReturn("001").`when`(projection).getItemNo()
        doReturn("기본").`when`(projection).getItemName()
        doReturn(1000L).`when`(projection).getItemPrice()
        doReturn(PRD001.ON_SALE.code).`when`(projection).getItemStatusCode()
        doReturn(10L).`when`(projection).getStock()
        doReturn(2L).`when`(projection).getQuantity()
        doReturn(false).`when`(projection).getIsOrder()
        doReturn(LocalDateTime.of(2026, 4, 28, 0, 0)).`when`(projection).getRegistDateTime()
        return projection
    }
}
